### PR TITLE
feat(ecs-patterns): create dlq when queue is not provided for QueueProcessingService

### PR DESCRIPTION
**Pull request by SeekerWing**
When a queue is not provided as input, create a queue as well as it's corresponding deadletterqueue(dlq), and define the redrive policy for when messages are sent to the dlq.

Not considered a breaking change as [RedrivePolicy](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-redrive) and [MessageRetentionPeriod](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html#aws-sqs-queue-msgretentionperiod) can be added with no interruption.

Fixes https://github.com/aws/aws-cdk/issues/6352

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs-patterns/README.md [hunk 1]
```diff
@@ -269,6 +269,8 @@ const queueProcessingFargateService = new QueueProcessingFargateService(stack, '
 });
 ```
 
+when queue not provided by user, CDK will create a primary queue and a dead letter queue with default redrive policy and attach permission to the task to be able to acces the primary queue.
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/README.md]: nit: typo access

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts [hunk 2]
```diff
@@ -191,8 +204,23 @@ export abstract class QueueProcessingServiceBase extends Construct {
...
     this.cluster = props.cluster || this.getDefaultCluster(this, props.vpc);
 
-    // Create the SQS queue if one is not provided
-    this.sqsQueue = props.queue !== undefined ? props.queue : new Queue(this, 'EcsProcessingQueue', {});
+    // Create the SQS queue if one is not provided, when creating SQS queue setup DLQ
+    if (props.queue) {
+      this.sqsQueue = props.queue;
+    } else {
+      this.deadLetterQueue = new Queue(this, "EcsProcessingDeadLetterQueue", {
+        retentionPeriod: Duration.days(14)
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: I would allow the retentionPeriod to be configured and 14 seems like a magic number to me. I think it would make more sense to allow sqs to default this to 4 days (do not specify a default, allow it to pass undefined), if the user does not specify it.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts [hunk 3]
```diff
@@ -191,8 +204,23 @@ export abstract class QueueProcessingServiceBase extends Construct {
...
+      });
+      this.sqsQueue = new Queue(this, 'EcsProcessingQueue', {
+        deadLetterQueue: {
+          queue: this.deadLetterQueue,
+          maxReceiveCount: props.maxReceiveCount || 5
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: Curious as to why you decided to default to 5 (seems like a magic number). If this is not a mandatory property, I would lean towards not defining a default, but rather only propagating the property if it's passed in.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts [hunk 4]
```diff
@@ -191,8 +204,23 @@ export abstract class QueueProcessingServiceBase extends Construct {
     }
     this.cluster = props.cluster || this.getDefaultCluster(this, props.vpc);
 
-    // Create the SQS queue if one is not provided
-    this.sqsQueue = props.queue !== undefined ? props.queue : new Queue(this, 'EcsProcessingQueue', {});
+    // Create the SQS queue if one is not provided, when creating SQS queue setup DLQ
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: You might want to rephrase this as `Create the SQS queue and it's corresponding DLQ if one is not provided`

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts [hunk 5]
```diff
@@ -143,6 +151,11 @@ export abstract class QueueProcessingServiceBase extends Construct {
    */
   public readonly sqsQueue: IQueue;
 
+  /**
+   * The dead letter queue for the SQS queue that the service will process from
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: Not sure if it makes sense to say the service will process from, might want to rephrase this as `The dead letter queue for the SQS queue`

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts [hunk 6]
```diff
@@ -86,6 +86,14 @@ export interface QueueProcessingServiceBaseProps {
    */
   readonly queue?: IQueue;
 
+  /**
+   * The maximum number of times that a message can be received by consumers.
+   * When this value is exceeded for a message the message will be automatically sent to the Dead Letter Queue.
+   *
+   * @default 5
```
Comment (pkandasamy91) [inline packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: Curious as to why you default to 5?

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts (hunk 3 above)
Comment (SeekerWing) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: i don't think maxReceiveCount is an optional attribute https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-sqs/lib/queue.ts#L142-L155

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts (hunk 2 above)
Comment (SeekerWing) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: it's not a magic number, it's in fact the max retention period for a DLQ. and i think this is a fair assumption to make for a DLQ, we want to keep it for the longest possible period.  

reference to CDKs affinity to this thought process https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/function.ts#L687-L689

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts (hunk 6 above)
Comment (SeekerWing) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: it is a magic number, it is however configurable by user.

File: packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts (hunk 2 above)
Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: Can we make this configurable?

Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: Can you make this configurable (as an input property)?

Comment (SeekerWing) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: i believe the lesser the configs an user has to learn and understand the better; i don't see the downsides of setting it to max value. my reasons for believing so:

* user doesn't have to pay more
* user has message for max possible time in order to deal with it
* if user is really proactive about dealing with dlq messages they'll consume and delete sooner than 14 days

Comment (pkandasamy91) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: As a user I should have the freedom to change any property within the construct. Otherwise, a user would need to re-write the entire construct in order to define a different value. 

A use case would be if an organization had an internal policy for how long specific data can be retained (7 days for instance), this 14 day default would prevent users from abiding by that policy, but by making it configurable, a user can override this default.

Comment (SeekerWing) [reply packages/@aws-cdk/aws-ecs-patterns/lib/base/queue-processing-service-base.ts]: make sense.
